### ISSUE TITLE
I shall open an Issue - Error tk Mac OSX

### ISSUE THREAD

**Issue by tcptps** [dynobo/normcap#97]
It said so so I did.
Mac OS Sierra (10.12)
installed with pip (Python3.9)
launched from Terminal normcap 
useful Application Window never appeared. 

If you need anymore information ask for it I am happy to help. 

<img width="856" alt="Bildschirmfoto 2021-03-05 um 11 42 11" src="https://user-images.githubusercontent.com/50037207/110106107-8dfc4f80-7da9-11eb-9237-9f00aa912416.png">

**Comment by dynobo** [comment]
Thanks for the report, @tcptps ! :+1: 
I actually didn't test on MacOS for quite a while (I don't have a Mac myself), but I try to take a look at it next weekend.

**Comment by dynobo** [comment]
Hi @tcptps,

today I installed Sierra in a VM and tried to reproduce the issue, but without success.

Can you please provide additional information?
1. Do you have a multi-monitor setup or only one display?
2. You wrote "useful Application Window never appeared.". Are you sure? The "application window" is barely visible, because it shows only a screenshot from your current desktop. The red border it draws is basically the only thing visible:
![2021-03-20_18-25](https://user-images.githubusercontent.com/11071876/111880121-c4d98480-89a9-11eb-8c1c-1c43e2fe9aa2.png)
If you see the red border, try selecting a portion of the screen with text to trigger the recognition process.
3. I'm not so familiar with MacOS, but I know there are some options how application windows show up. I think you can configure Mac to always maximize windows or something. Do you have something like that configure?

Thanks!

**Comment by tcptps** [comment]
Hi,
1. one Display 
2. No red border for me. 
3. Thats a weird one I have to say, because the Menu bar is hidden which is normally a sign that an application is in full Screen mode but on the other hand there is the little window. Using python 3.9 I think I used the Installer from the python website.
Since my Setup is weird sometimes and it works in your VM I would suggest to wait if somebody else experiences the same Problem and assume my Setup is just broken somehow. If there is any test I can run let me know. I tried using the development Environment: https://github.com/dynobo/normcap#setup-environment  but no luck either: 
<img width="1680" alt="Bildschirmfoto 2021-03-20 um 20 31 56" src="https://user-images.githubusercontent.com/50037207/111883596-a335c880-89bc-11eb-842a-60556360ba75.png">

**Comment by dynobo** [comment]
You observations are quite interesting!

What technically should happen is, that `normcap` opens two windows: 
1. [A "dummy" window](https://github.com/dynobo/normcap/blob/39990acd4cc16581a7f4fc2111980b0557457a93/normcap/handlers/crop_handler.py#L144), which seems to be necessary to receive keyboard shortcuts. This is the small window you see with and prompts you to report the bug.
2. Window 1) should never be visible, because a [child window](https://github.com/dynobo/normcap/blob/39990acd4cc16581a7f4fc2111980b0557457a93/normcap/handlers/crop_handler.py#L220) should open fullscreen in front of it showing the screenshot with red border.

Setting window 2) to fullscreen was quite difficult, because the behavior is different across the operation systems and because I also wanted multi-monitor setup. Especially on MacOS it was a struggle. The best I could come up with so far is [a certain combination of options](https://github.com/dynobo/normcap/blob/39990acd4cc16581a7f4fc2111980b0557457a93/normcap/handlers/crop_handler.py#L235).

Now the interesting part for me: 
While I read, that the options I use for Mac _should_ result in a fullscreen and topmost window, in my tests  it never _really_ was topmost: menubar and dock always were in front of it. I thought this might be the MacOS way, not allowing a real top-most window. In _your_ screenshot on the other hand, it seems like menubar and dock are somehow hidden, which is quite interesting. 

It looks a bit like in your case fullscreen mode is somehow enabled, but window 2) just not visible...

In any case, I'd be happy about help or hints anyone.

**Comment by dynobo** [comment]
I tried a change which _might_ fix this problem. Could anyone affected by this issue please try v0.1.11 and report back?
```
pip install normcap==0.1.11
```

**Comment by 9001** [comment]
just bumped into your project through mss, cool stuff :>  
noticed this issue and gave it a try since i had a mac nearby.

encountered somewhat similar behavior on 0.1.10:
* the red border is painted on startup, however the menu bar is on a layer above the border and renders over it
* capturing contents inside the menu bar is impossible; normcap reports that nothing was captured
* the dummy window appears when initiating the selection (click/drag)

can confirm that 0.1.11 mostly fixes both of these in my case;
* the red border is still covered by the menu bar until the left mouse button is clicked, this immediately makes the border top-level
* ...so capturing contents inside the menu bar is now possible by starting a selection right below it
* the dummy window no longer appears at all

this was using the system-provided python 3.8 and tkinter on catalina 10.15.7

**Comment by dynobo** [comment]
@9001 Thanks for your valuable feedback! This seems like updating to `v0.1.11` is a good workaround for people on MacOS affected by this issue,

I leave this ticket open, because it still doesn't behave as it should (menu bar should be covered, too.) 

I'm currently experimenting with a different GUI-Framework ([Kivy](https://kivy.org/) that hopefully fixes this issue and makes other things more easier for me. I'll drop a notice here, as soon as it's ready for testing.

**Comment by y4rr** [comment]
@dynobo I'm not a super expert in developing for MacOS, but I think I may have here something interesting for you. Here's an example of how to use PyObjC to draw a window which in the fullscreen mode will render over the menu bar and the dock:

```
import objc
from Foundation import NSObject
from AppKit import NSApp, NSApplication, NSWindow, NSScreen, NSNormalWindowLevel, NSFloatingWindowLevel, NSTitledWindowMask, NSResizableWindowMask, NSClosableWindowMask, NSBorderlessWindowMask, NSBackingStoreBuffered, NSWindowCollectionBehaviorFullScreenPrimary
from PyObjCTools import AppHelper


class MyWindow(NSObject):
    def __new__(cls, *args, **kwargs):
        return cls.alloc().init()

    def __init__(self):
        frame = ((0.0, 0.0), (640.0, 480.0))
        mask = NSTitledWindowMask | NSResizableWindowMask | NSClosableWindowMask | NSBorderlessWindowMask
        screen = NSScreen.mainScreen()
        #level = NSFloatingWindowLevel
        # It's just an integer, you can set it for something ridiculously high, so for example in the full screen mode the window will render over the menu bar and the dock
        level = 999999
        self._window = NSWindow.alloc().initWithContentRect_styleMask_backing_defer_screen_(frame, mask, NSBackingStoreBuffered, False, screen)
        self._window.setTitle_("Hello World")
        self._window.setLevel_(level)
        self._window.setDelegate_(self)
        self._window.setCollectionBehavior_(NSWindowCollectionBehaviorFullScreenPrimary)
        self._window.makeKeyAndOrderFront_(self)
        self._window.retain()


def main():
    app = NSApplication.sharedApplication()

    win = MyWindow()

    NSApp.activateIgnoringOtherApps_(True)

    AppHelper.runEventLoop()

if __name__ == "__main__":
    main()
```

I hope this might be useful for this issue.

**Comment by dynobo** [comment]
@y4rr Thanks a lot for your efforts and this information. I'm currently working hard on a switch from Tk to Qt (see the respective branch), which hopefully makes it more easy to maintain NormCap for multiple platforms. Interestingly, showing the fullscreen already works quite well with Qt on MacOS, but there are some other Mac-specific issues I have to sort out, e.g. I can't change the cursor to crosshair yet and the window doesn't get focused to receive keypress events... But maybe objc can be of use here, too! :-)

**Comment by dynobo** [comment]
In the new [beta release 0.2.0-b1](https://github.com/dynobo/normcap/releases/tag/v0.2.0-b1) I got completely rid of the "dummy window". In general it should provide a much nicer experience on MacOS (still not perfect, e.g. it still is behind the application menu and dock).

Please help by testing and [providing feedback](https://github.com/dynobo/normcap/issues/110).

**Comment by dynobo** [comment]
The issue with "dummy window" is solved with #116 and has been published in the new v0.2.0b3.

For the problem that the window stay below the menubar, I created Issue #119
